### ISSUE TITLE
Python 3.12 on Ubuntu 24.04 is Externally Managed - PIP is broken

### ISSUE THREAD

**Issue by molson504x** [actions/runner-images#10781]
### Description

Python 3.12 on the `ubuntu-24.04` image (which is also slowly becoming `ubuntu-latest`) is using an ["Externally Managed" instance](https://packaging.python.org/en/latest/specifications/externally-managed-environments/) of Python - either it was installed via a package manager (apt?) or it is packaged with the OS.  This also uses PIP 24, which (starting with PIP 23) [enforces EXTERNALLY-MANAGED flagged python installs](https://discuss.python.org/t/announcement-pip-23-0-release/23342). 

If your workflow includes anything that tries to do a `pip install` to set up python dependencies, it will fail with an error similar to:

```
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try apt install
    python3-xyz, where xyz is the package you are trying to
    install.
    
    If you wish to install a non-Debian-packaged Python package,
    create a virtual environment using python3 -m venv path/to/venv.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
    sure you have python3-full installed.
    
    If you wish to install a non-Debian packaged Python application,
    it may be easiest to use pipx install xyz, which will manage a
    virtual environment for you. Make sure you have pipx installed.
    
    See /usr/share/doc/python3.12/README.venv for more information.
```

We found this by accident because our workflows target `ubuntu-latest` which is slowly being changed to target `ubuntu-24.04` instead of `ubuntu-22.04`.  We confirmed that adding a `setup-python` step to our workflow (which sets up 3.12.3) fixes the problem because it'll install a stand-alone interpreter and update the system path to use that interpreter, however there is no documentation anywhere saying that is now a requirement.  Hence this issue :) 

### Platforms affected

- [ ] Azure DevOps
- [X] GitHub Actions - Standard Runners
- [ ] GitHub Actions - Larger Runners

### Runner images affected

- [ ] Ubuntu 20.04
- [ ] Ubuntu 22.04
- [X] Ubuntu 24.04
- [ ] macOS 12
- [ ] macOS 13
- [ ] macOS 13 Arm64
- [ ] macOS 14
- [ ] macOS 14 Arm64
- [ ] macOS 15
- [ ] macOS 15 Arm64
- [ ] Windows Server 2019
- [ ] Windows Server 2022

### Image version and build link

```
  Image: ubuntu-24.04
  Version: 20241006.1.0
```

HOWEVER, we were targeting `ubuntu-latest` and discovered on subsequent runs that `ubuntu-latest` targeted `ubuntu-22.04` which made this intermittent for us.  When we hard-targeted `ubuntu-24.04` we consistently reproduced this.

I can't provide a build link due to client confidentiality purposes, however this is pretty easy to test.

### Is it regression?

Yes (since the behavior of ubuntu-latest changed and broke a workflow)

### Expected behavior

PIP should continue to work, or GitHub needs to formally publish the breaking change along with remediation steps (which is to use the setup-python action).  Ideally, the version of Python that comes pre-installed on the runner image is not externally managed.

### Actual behavior

See the error in the description... :) 

Also - the inconsistency with using `ubuntu-latest` being either `ubuntu-22.04` and `ubuntu-24.04` is really bad.  This is going to cause a lot of painful debugging for a lot of people because of this inconsistency causing intermediate but not repeatable failures.

### Repro steps

1. Prepare a workflow which tries to install a python dependency using PIP and targets `ubuntu-24.04`.
2. Observe the error.

REMEDIATION:
1. Prepare a workflow with steps to install python (`setup-python`) and then tries to use PIP.  This workflow should target `ubuntu-24.04`.
2. Observe no error and the machines are all happy.

**Comment by molson504x** [comment]
> To unblock yourself, please consider using ubuntu-22.04 runner in your workflow.
> 
> `runs-on: ubuntu-22.04`
> 
> Cause : ubuntu-24.04 runner uses python 3.12 as the default installation, which has a marker file _EXTERNALLY-MANAGED_. This file insists and tolls like pip install packages within a virtual environment.

Hey @subir0071  - yeah, that was one of our solutions.  We actually did it a different way.  We added a `setup-python` step to the workflow, and since that installs a python interpreter that is not externally managed it solves the problem.  HOWEVER, my main issue with this is it's a breaking change which is not documented, and it is an intermittent breaking change due to the phased rollout of updating the `ubuntu-latest` tag.  This breaking change should really be documented somewhere.

_OR_ - even better would be to package the runner image with a python interpreter which is not externally managed.

**Comment by lpascal-ledger** [comment]
Hello @subir0071 :wave: 

Are you suggesting that the issue will not be managed in `ubuntu-24.04`, and the image, by default, will always trigger an error when doing a `pip install <something>` ?

**Comment by gtskaushik** [comment]
This broke our builds too

**Comment by hmpf** [comment]
We just ran into this ourselves.

The technically correct solution is to always use virtual envs, absolutely everywhere, if it isn't a distro package. Another solution is to install on the user with pipx, but that's not quite smooth sailing either: https://github.com/pypa/pipx/issues/982

**Comment by molson504x** [comment]
> We just ran into this ourselves.
> 
> The technically correct solution is to always use virtual envs, absolutely everywhere, if it isn't a distro package. Another solution is to install on the user with pipx, but that's not quite smooth sailing either: [pypa/pipx#982](https://github.com/pypa/pipx/issues/982)

No, the technically correct solution is to use a stand-alone python interpreter (using the `setup-python` step.  And no, `pipx` is not a good solution because, while it _might_ work it still exposes the risk of messing up the underlying OS functionality (which is what pip is trying to protect from by implementing the EXTERNALLY-MANAGED flag check).

> Hi @molson504x , Hope you are able to progress with your CI/CD workload. Kindly suggest on the closure of this ticket. Thanks.

Hey @subir0071  - Respectfully, no.  This issue is not resolved, it just has a couple of work-arounds to unblock people.  The root cause of the problem is that the image being used for the `ubuntu-24.04` runners is using the python interpreter which ships with the OS, which is externally managed.  GitHub is advertising out-of-the-box python support similar to 22.04, which is not the case here because pip (a core tool to python) won't allow installation of modules in this configuration.  The solution is that the image needs to ship with a stand-alone python interpreter installed so that it isn't trying to use the managed instance that ships with the underlying OS.  Work-arounds are fine, but saying "just use `ubuntu-22.04`" or "just use the `setup-python` step" or "just use a virtual environment" is a behavioral change from 22.04 which is not documented, and either needs to be resolved or documented before I will accept closing this issue.

**Comment by geekosaur** [comment]
I respectfully disagree, it is not incumbent on GitHub to fix issues with an upstream distribution. (We (haskell/cabal) have a different issue with testing, but we're not expecting GitHub to reintroduce packages that Canonical removed from 24.04.) Using an action or a venv to install a private/safe Python was always the correct thing to do, and Canonical as of 24.04 is enforcing it.

**Comment by liamhuber** [comment]
> ...either needs to be resolved or documented before I will accept closing this issue.

Agreed. The image (documentation page](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2404-Readme.md) lists:

> Language and Runtime
> - ...
> - Python 3.10.12
> - ...
> 
> Package Management
> - ...
> - Pip 24.0
> - ...

Without any notes, it feels misleading to me to have this listed yet `pip install...` immediately raises an error on the image.

I would find it most convenient to be able to continue usage as with 22.04, but a documentation update seems reasonable.

**Comment by srl295** [comment]
for what it's worth, `pipenv install` seemed to also trigger this issue. So, I thought I was using a virtual environment.

I was not able to reproduce this using an `ubuntu:24.04` docker image. (i.e. pipenv install succeeded). I'll probably use the `runs-on: ubuntu-22.04` instead.

**Comment by molson504x** [comment]
@subir0071 - Based on the post in #10636 I'm willing to close this, however this should still be addressed in documentation about `ubuntu-24.04` runners or should be addressed by installing a non-system-managed python interpreter in the image.

Comment, for reference:

> Hey everyone - thank you for your input. Based on community feedback we have decided to rollback the ubuntu-latest migration. All runs on ubuntu-latest will go back to Ubuntu-22 while we work out the issues with the image. We apologize for any service disruption you all may have had, and appreciate everyone’s feedback and patience. The rollback will be completed in the next hour.
> 
> To give some context for the changes we made - our images have become very large over the past 2 years and we were trying to free up some space to give us some breathing room going forward. There is currently no extra free space left for installing any more tooling, while still maintaining our free space SLA of 14gb. The size of the image also impacts performance of the VMs as well. Since a new OS is always a cutover, we took the opportunity to free up some space. Clearly this was not the right approach, and we apologize for the pain we have caused.
> 
> We will re-evaluate our communication and rollout strategy before beginning this migration again.

**Comment by molson504x** [comment]
Closing this issue, per comment in #10636 - rolling back `ubuntu-latest` to reference `ubuntu-22.04` runner image.

**Comment by geekosaur** [comment]
Yes, I know this has been closed, but I want to provide a few examples supporting my stance on this.

> Using an action or a venv to install a private/safe Python was always the correct thing to do, and Canonical as of 24.04 is enforcing it.

- Back in 2006-2007 a requested update to a Perl module in our environment broke (Open)SuSE's `yast`.
- Roughly a decade later, I tried to help someone repair their Ubuntu system after upgrading a Python module that broke `apt-get`. (Ultimately they ended up reinstalling; the broken module broke post-install scripts as well, so we couldn't even fall back to `dselect`/`dpkg`.)

Red Hat / Fedora avoid this by shipping their own minimal Python install that system tools use, avoiding the user-facing one, but I don't think this scales especially if you're trying to keep image sizes down. The _only_ reliable way to avoid it is to assume the system-provided language packages are intended to work with the system and install your own if you will be upgrading or installing additional packages not via the system's package manager.

**Comment by eli-schwartz** [comment]
> Using an action or a venv to install a private/safe Python was always the correct thing to do, and Canonical as of 24.04 is enforcing it.

This is both wrong and fallaciously wrong.

Inside of a temporary container, all the reasons for installing additional python interpreters just to have a different environment, are not applicable. The original rationale for this "Externally Managed" change even notes that scenario. But "Ubuntu Linux" doesn't know when it is being used "as a container", that's been deferred to container vendors such as Github Actions to fix by deleting the marker file. Of course, no container vendors bother to do so.

**Comment by srl295** [comment]
> > Using an action or a venv to install a private/safe Python was always the correct thing to do, and Canonical as of 24.04 is enforcing it.
> 
> This is both wrong and fallaciously wrong.
> 
> Inside of a temporary container, all the reasons for installing additional python interpreters just to have a different environment, are not applicable. The original rationale for this "Externally Managed" change even notes that scenario. But "Ubuntu Linux" doesn't know when it is being used "as a container", that's been deferred to container vendors such as Github Actions to fix by deleting the marker file. Of course, no container vendors bother to do so.

Specifically, at the end of [this section](https://packaging.python.org/en/latest/specifications/externally-managed-environments/#marking-an-interpreter-as-using-an-external-package-manager) of the Python docs:

> _In certain contexts, such as single-application container images that aren’t updated after creation, a distributor may choose not to ship an EXTERNALLY-MANAGED file, so that users can install whatever they like (as they can today) without having to manually override this rule._

However, the recommendation section [here](https://packaging.python.org/en/latest/specifications/externally-managed-environments/#keep-the-marker-file-in-container-images) recommends to

> _Keep the marker file in container images_

**Comment by janbrasna** [comment]
Just for visibility here, this issue will be coming back:
- #10897

**Comment by bmaehr** [comment]
Just as are reverence: You get the same error (because pip is ignoring your virtual environment) if the directory for your virtual environment is named wrong e.g. `venv` instead of `.venv`.
